### PR TITLE
[CUDA] Fix out-of-bounds/invalid-gradient in layer_norm's huge-M gamma/beta backward

### PR DESCRIPTION
**Pull request by anatoliylitv**
## Summary

`LaunchGammaBetaBackwardCUDAKernel`'s huge-M path (`M > 64 * 1024 && N / warp_size < sm_count / 2`) sizes its `dgamma_blocks` / `dbeta_blocks` scratch buffers from `dgamma->size(-1)` instead of `N`. This code sits outside every `#ifdef USE_ROCM`, so it's a shared CUDA/ROCm bug, not ROCm-specific.

`dgamma->size(-1)` is wrong in two cases:
- **Multi-dimensional `normalized_shape`**: `size(-1)` is only the last dim, which can be narrower than `N` (the product of all normalized dims). The kernel writes `N` columns per row into a buffer allocated with too few columns, an out-of-bounds device write.
- **Undefined gamma (`weight=None`)**: `size(-1)` on an undefined tensor silently returns `0` instead of raising, so `dbeta_blocks` (which is otherwise legitimately needed, since `bias` can still be defined) is allocated empty.

Both are reachable on CUDA as well as ROCm, any time `M > 64 * 1024` and `N` is small relative to `sm_count`.

## Reproducer

[layer_norm_huge_m_buffer_bug_reproducer.py](https://github.com/user-attachments/files/31473546/layer_norm_huge_m_buffer_bug_reproducer.py)

```
=== multidim_oob ===
  F.layer_norm(x[100000,8,16], (8,16), w, b).sum().backward() -> out-of-bounds device write / GPU memory fault
  -> BUG: backward() raised: Function NativeLayerNormBackward0 returned an invalid gradient at index 1 - got [16] but expected shape compatible with [8, 16]

=== weight_none_invalid_grad ===
  F.layer_norm(x[100000,64], [64], None, bias).sum().backward() -> invalid gradient shape [0] vs [64]
  -> BUG: backward() raised: Function NativeLayerNormBackward0 returned an invalid gradient at index 2 - got [0] but expected shape compatible with [64]
```

## Fix

Allocate `dgamma_blocks` / `dbeta_blocks` with `N` columns (the width `GammaBetaBackwardCUDAKernelTemplate` actually writes via `dg[thread_y * N + thread_x]`) instead of `dgamma->size(-1)` / an assumed-equal `dbeta` width, using each tensor's own `.options()`. Since the final `.sum(0)` now produces a flat `{N}` tensor that may not match the gradient's own (possibly multi-dim) shape, `.view_as(*dgamma)` / `.view_as(*dbeta)` reshape it back before assignment.

Reproducer result:
```
=== multidim_oob ===
  F.layer_norm(x[100000,8,16], (8,16), w, b).sum().backward() -> out-of-bounds device write / GPU memory fault
  -> FIXED: weight.grad/bias.grad have correct shape and match the CPU float32 reference

=== weight_none_invalid_grad ===
  F.layer_norm(x[100000,64], [64], None, bias).sum().backward() -> invalid gradient shape [0] vs [64]
  -> FIXED: bias.grad has correct shape and matches the CPU float32 reference
```

## Tests

Added `test_layer_norm_gamma_beta_backward_huge_M_buffers` to `test/test_nn.py` (`TestNNDeviceType`), parametrized over `op,normalized_shape,has_weight` at `M=100000` to exercise the huge-M path, with four named cases:

- `layer_norm_1d_no_weight`: 1-D `normalized_shape=[64]`, `weight=None`, `bias.requires_grad=True` — covers the invalid-gradient repro.
- `layer_norm_multidim`: multi-dim `normalized_shape=(8, 16)`, weight and bias both defined — covers the out-of-bounds-write repro.
- `layer_norm_multidim_no_weight`: multi-dim `normalized_shape=(8, 16)`, `weight=None` — both failure modes at once.
- `rms_norm_multidim`: multi-dim `normalized_shape=(8, 16)`, weight defined — `RMSNormBackwardKernelImpl` routes into the same fixed allocation. (`weight=None` is not meaningful for `rms_norm`: no gamma gradient at all in that case.)

Used AI assistance from Cursor.

### CODE REVIEW

**Review by jeffdaily — CHANGES_REQUESTED** [review]
The diagnosis and fix look correct: the kernel writes N columns per row (`dg[thread_y * N + thread_x]`), so sizing the scratch buffers from `dgamma->size(-1)` was wrong for multi-dim `normalized_shape` and for undefined gamma. Three things before this can land:

1. **Rebase needed.** The branch conflicts with main. This region was refactored recently (the huge-M condition moved into `ShouldUseHugeMGammaBetaBackwardKernel`, and the caller grew a ROCm two-pass dispatch). The buggy allocations are still present on main, so your fix carries over cleanly, but please rebase or merge main first.

2. **Comment fix.** The new comment says the old code borrowed dgamma's options for `dbeta_blocks`. It didn't; it already used `dbeta->options()`, only the width came from `dgamma->size(-1)`. Please correct that, and consider shortening both comments to just the invariant, something like: "The kernel writes N columns per row; dgamma->size(-1) is only the last normalized dim and is 0 for an undefined tensor."

3. **rms_norm test.** The fixed allocation is shared by the rms_norm path (`RMSNormBackwardKernelImpl` routes into the same launcher), and `F.rms_norm` with multi-dim `normalized_shape` at huge M hit the same out-of-bounds write. Please add a huge-M multi-dim rms_norm case, or parametrize the multidim test over layer_norm/rms_norm.

Minor: the huge-M path is only taken when `N / warp_size < sm_count / 2`. A one-line note in the multidim test about that assumption would make it visible if a small-GPU runner ever stops exercising the path.

The two layer_norm tests are good coverage; they hit corners the existing `test_layer_norm_backwards_eps` can't reach (weight=None with bias, multi-dim normalized_shape at huge M).

This review was drafted with AI assistance (Claude).

**Comment by anatoliylitv** [comment]
Mentioned in description test result after fix:
```
=== multidim_oob ===
  F.layer_norm(x[100000,8,16], (8,16), w, b).sum().backward() -> out-of-bounds device write / GPU memory fault
  -> FIXED: weight.grad/bias.grad have correct shape and match the CPU float32 reference

=== weight_none_invalid_grad ===
  F.layer_norm(x[100000,64], [64], None, bias).sum().backward() -> invalid gradient shape [0] vs [64]
  -> FIXED: bias.grad has correct shape and matches the CPU float32 reference
```

File: test/test_nn.py [hunk 1]
```diff
@@ -8279,6 +8279,96 @@ def test_layer_norm_backward_undefined_gamma(self, device):
 
         self.assertEqual(bias.grad.cpu(), bias_cpu.grad, f"M={M} N={N}", atol=1e-4, rtol=1e-4)
 
+    @onlyCUDA
+    def test_layer_norm_backward_huge_M_undefined_gamma(self, device):
```
Comment (eqy) [inline test/test_nn.py]: Can these tests be folded together into one?

Comment (anatoliylitv) [reply test/test_nn.py]: Merged new unit tests together.

**Review by jeffdaily — APPROVED** [review]
LGTM. All four points from my earlier review were addressed in the 2026-08-18 push, along with eqy's request to fold the tests into one parametrized test.

I confirmed the rms_norm subtest reaches the patched code rather than the composite (`_fused_rms_norm` -> `_fused_rms_norm_backward` -> the huge-M branch), and pulled the test-report XMLs from run 32136173109 to confirm all four subtests actually ran and passed on the gfx950 shards.

I pushed two maintainer commits: a test fix plus a merge of the latest viable/strict. The test built weight and bias on CPU and then did `weight.to(dev).requires_grad_(True)` in a loop over `(device, "cpu")`. On the "cpu" iteration `.to("cpu")` returns the same object, so `requires_grad_` flipped the flag on the shared source tensor in place. It worked only because the CUDA iteration ran first; reversing the tuple makes `w` a non-leaf, `w.grad` None, and the comparison loop dies on `grad.shape`. Verified both orders on gfx90a.

Two non-blocking suggestions for next time. The test hardcodes `M = 100000` and never checks the huge-M path was selected, so a retuned threshold would leave all four subtests passing while covering the ordinary tiled path; the neighbouring two-pass test derives its shape from device properties instead. And `view_as` is needed only because this branch replaces the caller's tensor, `at::sum_out` into `dgamma->view({N})` would drop both the wasted allocation and the reshape.

Silent wrong-gradient and an out-of-bounds device write on stock CUDA, not just ROCm, so worth a `topic: bug fix` label and a release-branch cherry-pick.

Analysis assisted by AI.
